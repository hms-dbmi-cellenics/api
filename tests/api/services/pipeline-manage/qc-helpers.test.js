
const { getQcStepsToRun } = require('../../../../src/api/services/pipelines/manage/getQcStepsToRun');
const { buildQCPipelineSteps } = require('../../../../src/api/services/pipelines/manage/skeletons/buildQCPipelineSteps');
const fake = require('../../../test-utils/constants');

jest.mock('../../../../src/utils/aws/s3', () => ({
  fileExists: jest.fn(() => true),
}));

describe('helper functions for skeletons', () => {
  it('returns the correct first step given a list', async () => {
    const processingConfig = [
      {
        name: 'numGenesVsNumUmis',
        body: {
          auto: true,
          filterSettings: {
            regressionType: 'linear',
            regressionTypeSettings: {
              linear: {
                'p.level': 0.001,
              },
              spline: {
                'p.level': 0.001,
              },
            },
          },
          enabled: true,
          '8e6ffc70-14c1-425f-b1be-cef9656a55a5': {
            auto: true,
            filterSettings: {
              regressionType: 'linear',
              regressionTypeSettings: {
                linear: {
                  'p.level': 0.0002898551,
                },
                spline: {
                  'p.level': 0.001,
                },
              },
            },
            defaultFilterSettings: {
              regressionType: 'linear',
              regressionTypeSettings: {
                linear: {
                  'p.level': 0.0002898551,
                },
                spline: {
                  'p.level': 0.001,
                },
              },
            },
            api_url: 'http://host.docker.internal:3000',
            enabled: true,
          },
        },
      },
      {
        name: 'cellSizeDistribution',
        body: {
          auto: true,
          filterSettings: {
            minCellSize: 1080,
            binStep: 200,
          },
          enabled: false,
          '8e6ffc70-14c1-425f-b1be-cef9656a55a5': {
            auto: true,
            filterSettings: {
              minCellSize: 1136,
              binStep: 200,
            },
            defaultFilterSettings: {
              minCellSize: 1136,
              binStep: 200,
            },
            api_url: 'http://host.docker.internal:3000',
            enabled: false,
          },
        },
      },
      {
        name: 'doubletScores',
        body: {
          auto: true,
          filterSettings: {
            probabilityThreshold: 0.5,
            binStep: 0.05,
          },
          enabled: true,
          '8e6ffc70-14c1-425f-b1be-cef9656a55a5': {
            auto: true,
            filterSettings: {
              probabilityThreshold: 0.6506245,
              binStep: 0.05,
            },
            defaultFilterSettings: {
              probabilityThreshold: 0.6506245,
              binStep: 0.05,
            },
            api_url: 'http://host.docker.internal:3000',
            enabled: true,
          },
        },
      },
    ];

    const qcSteps = await getQcStepsToRun(fake.EXPERIMENT_ID, processingConfig);
    expect(qcSteps[0]).toEqual('CellSizeDistributionFilterMap');
  });

  it('returns the default first step and full state machine if the config has no updates', async () => {
    const processingConfig = [];

    const qcSteps = await getQcStepsToRun(fake.EXPERIMENT_ID, processingConfig);
    expect(qcSteps[0]).toEqual('ClassifierFilterMap');
    const stateMachine = buildQCPipelineSteps(qcSteps);
    expect(stateMachine).toMatchSnapshot();
  });
});
