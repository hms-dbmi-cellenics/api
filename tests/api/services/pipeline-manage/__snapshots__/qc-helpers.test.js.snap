// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper functions for skeletons returns the default first step and full state machine if the config has no updates 1`] = `
Object {
  "CellSizeDistributionFilterMap": Object {
    "ItemsPath": "$.samples",
    "Iterator": Object {
      "StartAt": "CellSizeDistributionFilter",
      "States": Object {
        "CellSizeDistributionFilter": Object {
          "End": true,
          "XConstructorArgs": Object {
            "perSample": true,
            "taskName": "cellSizeDistribution",
          },
          "XStepType": "create-new-step",
        },
      },
    },
    "Next": "MitochondrialContentFilterMap",
    "ResultPath": null,
    "Type": "Map",
  },
  "ClassifierFilterMap": Object {
    "ItemsPath": "$.samples",
    "Iterator": Object {
      "StartAt": "ClassifierFilter",
      "States": Object {
        "ClassifierFilter": Object {
          "End": true,
          "XConstructorArgs": Object {
            "perSample": true,
            "taskName": "classifier",
          },
          "XStepType": "create-new-step",
        },
      },
    },
    "Next": "CellSizeDistributionFilterMap",
    "ResultPath": null,
    "Type": "Map",
  },
  "ConfigureEmbedding": Object {
    "Next": "EndOfPipeline",
    "XConstructorArgs": Object {
      "perSample": false,
      "taskName": "configureEmbedding",
    },
    "XStepType": "create-new-step",
  },
  "DataIntegration": Object {
    "Next": "ConfigureEmbedding",
    "XConstructorArgs": Object {
      "perSample": false,
      "taskName": "dataIntegration",
      "uploadCountMatrix": true,
    },
    "XStepType": "create-new-step",
  },
  "DoubletScoresFilterMap": Object {
    "ItemsPath": "$.samples",
    "Iterator": Object {
      "StartAt": "DoubletScoresFilter",
      "States": Object {
        "DoubletScoresFilter": Object {
          "End": true,
          "XConstructorArgs": Object {
            "perSample": true,
            "taskName": "doubletScores",
          },
          "XStepType": "create-new-step",
        },
      },
    },
    "Next": "DataIntegration",
    "ResultPath": null,
    "Type": "Map",
  },
  "EndOfPipeline": Object {
    "End": true,
    "Type": "Pass",
  },
  "MitochondrialContentFilterMap": Object {
    "ItemsPath": "$.samples",
    "Iterator": Object {
      "StartAt": "MitochondrialContentFilter",
      "States": Object {
        "MitochondrialContentFilter": Object {
          "End": true,
          "XConstructorArgs": Object {
            "perSample": true,
            "taskName": "mitochondrialContent",
          },
          "XStepType": "create-new-step",
        },
      },
    },
    "Next": "NumGenesVsNumUmisFilterMap",
    "ResultPath": null,
    "Type": "Map",
  },
  "NumGenesVsNumUmisFilterMap": Object {
    "ItemsPath": "$.samples",
    "Iterator": Object {
      "StartAt": "NumGenesVsNumUmisFilter",
      "States": Object {
        "NumGenesVsNumUmisFilter": Object {
          "End": true,
          "XConstructorArgs": Object {
            "perSample": true,
            "taskName": "numGenesVsNumUmis",
          },
          "XStepType": "create-new-step",
        },
      },
    },
    "Next": "DoubletScoresFilterMap",
    "ResultPath": null,
    "Type": "Map",
  },
}
`;
